Sources/StreamChat/APIClient/DefaultCDNClient_Tests.swift

```diff
@@ -0,0 +1,223 @@
+
+        // We have to compare error codes, since iOS14 wraps network errors into `NSURLError`
+        // in which we cannot retrieve the wrapper error
+        XCTAssertEqual((decoder.decodeRequestResponse_error as NSError?)?.code, networkError.code)
+
+        // Check the outgoing data is from the decoder
+        XCTAssertEqual(result.error as? TestError, encoderError)
+    }
+
+    func test_callingUploadFile_createsNetworkRequest() throws {
+        let builder = Builder()
+        let client = builder.make()
+        
+        let attachment = ChatMessageAttachment.sample(
+            uploadingState: .init(
+                localFileURL: .sampleYodaImage,
+                state: .pendingUpload
+            )
+        )
+        
+        let uploadingState = try XCTUnwrap(attachment.uploadingState)
+        
+        let multipartFormData = MultipartFormData(
+            try Data(contentsOf: .sampleYodaImage),
+            fileName: uploadingState.localFileURL.lastPathComponent,
+            mimeType: uploadingState.localFileURL.attachmentFile?.type.mimeType
+        )
+
+        let uniquePath: String = .unique
+        let uniqueQueryItem: String = .unique
+        var testRequest = URLRequest(url: URL(string: "test://test.test/\(uniquePath)?item=\(uniqueQueryItem)")!)
+        testRequest.httpMethod = "post"
+        builder.encoder.encodeRequest = .success(testRequest)
+
+        // Simulate file uploading.
+        client.uploadAttachment(
+            .sample(
+                uploadingState: .init(
+                    localFileURL: .sampleYodaImage,
+                    state: .pendingUpload
+                )
+            ),
+            progress: nil,
+            completion: { _ in }
+        )
+
+        // Check a network request is made with the values from `testRequest`
+        AssertNetworkRequest(
+            method: .post,
+            path: "/" + uniquePath,
+            headers: [
+                "Content-Type": "multipart/form-data; boundary=\(MultipartFormData.boundary)"
+            ],
+            queryParameters: ["item": uniqueQueryItem],
+            body: multipartFormData.getMultipartFormData()
+        )
+    }
+}
+
+private class Builder {
+    let encoder = TestRequestEncoder(
+        baseURL: .unique(),
+        apiKey: .init(.unique)
+    )
+    let decoder = TestRequestDecoder()
+    var sessionConfiguration = URLSessionConfiguration.ephemeral
+    let uniqueHeaderValue = String.unique
+    
+    func make() -> DefaultCDNClient {
+        sessionConfiguration.httpAdditionalHeaders?["unique_value"] = uniqueHeaderValue
+        RequestRecorderURLProtocol.startTestSession(with: &sessionConfiguration)
+        MockNetworkURLProtocol.startTestSession(with: &sessionConfiguration)
+        
+        return DefaultCDNClient(
+            encoder: encoder,
+            decoder: decoder,
+            sessionConfiguration: sessionConfiguration
+        )
+    }
+}
```

Comment:
There is two ideas behind that:
- setUp() has an issue of shared state, ideally every test should be isolated
- it has no means of behaviour parametrisation which this newly introduced pattern helps with

As for your concern on two extra lines - Voijta actually has a pretty nice idea on how to optimise it, but even in the current state imho I don't see any serious issues with a couple of extra lines because the benefits outweigh the downsides and it's not like there are options here because shared fixture is a code smell in tests. So it's just a fundamental feature of any test - it has a setup stage